internal/pkg/agent/application/coordinator/coordinator.go

```diff
@@ -1216,10 +1219,103 @@ func (c *Coordinator) generateComponentModel() (err error) {
 	// If we made it this far, update our internal derived values and
 	// return with no error
 	c.derivedConfig = cfg
+	c.lastComponentModel = c.componentModel
 	c.componentModel = comps
+
+	// don't run a bunch of relatively costly code unless we're in debug mode
+	if c.logger.IsDebug() {
+		c.checkAndLogUpdate()
+	}
+
 	return nil
 }
 
+// compares the last component model with an updated model,
+// logging any differences.
+func (c *Coordinator) checkAndLogUpdate() {
+	if c.lastComponentModel == nil {
+		c.logger.Debugf("Received initial component update; total of %d components", len(c.componentModel))
+		return
+	}
+	type compCheck struct {
+		inNew  bool
+		inLast bool
+	}
+
+	type compOutputs struct {
+		inNew  bool
+		inLast bool
+	}
+
+	compOutputTypes := map[string]compOutputs{}
+	// collect current and last component list
+	compMap := map[string]compCheck{}
+	currentCompIDs := make([]string, len(c.componentModel))
+
+	// combine last and current lists
+	for i, newComp := range c.componentModel {
+		currentCompIDs[i] = newComp.ID
+		compMap[newComp.ID] = compCheck{inNew: true}
+		compOutputTypes[newComp.OutputType] = compOutputs{inNew: true}
+	}
+
+	for _, lastComp := range c.lastComponentModel {
+		if current, ok := compMap[lastComp.ID]; ok {
+			current.inLast = true
+			compMap[lastComp.ID] = current
+		} else {
+			compMap[lastComp.ID] = compCheck{inLast: true}
+		}
+		if current, ok := compOutputTypes[lastComp.OutputType]; ok {
+			current.inLast = true
+			compOutputTypes[lastComp.OutputType] = current
+		} else {
+			compOutputTypes[lastComp.OutputType] = compOutputs{inLast: true}
+		}
+	}
+
+	// reduce to list of added/removed inputs
+	addedList := []string{}
+	removedList := []string{}
+	for ID, comp := range compMap {
+		if comp.inLast && !comp.inNew {
+			removedList = append(removedList, ID)
+		}
+		if !comp.inLast && comp.inNew {
+			addedList = append(addedList, ID)
+		}
+
+	}
+
+	// reduced to list of added/removed outputs
+	addedOutputs := []string{}
+	removedOutputs := []string{}
+	for output, comp := range compOutputTypes {
+		if comp.inLast && !comp.inNew {
+			removedOutputs = append(removedOutputs, output)
+		}
+		if !comp.inLast && comp.inNew {
+			addedOutputs = append(addedOutputs, output)
+		}
+	}
+
+	if len(addedList) > 0 {
+		c.logger.Debugf("The following components have been added: %v", addedList)
+	}
+	if len(removedList) > 0 {
+		c.logger.Debugf("The following components have been removed: %v", removedList)
+	}
+
+	if len(addedOutputs) > 0 {
+		c.logger.Debugf("The following outputs have been added: %v", addedOutputs)
+	}
+	if len(removedOutputs) > 0 {
+		c.logger.Debugf("The following outputs have been removed: %v", removedOutputs)
+	}
```

Comment:
This doesn't log components that have been updated, so that's the biggest thing to add. It is probably the most complicated to compute though.

Reply:
Yah, that's a fairly complex operation, since it involves diffing two relatively complex objects, though.